content/blog/2022/2023-05-07-organize-code-in-nodejs.md

```diff
@@ -0,0 +1,295 @@
+        return res.status(404).send('Order not found')
+    }
+    res.json(order)
+})
+
+module.exports = {getOrders,getOrderByID,getOrderStatus}
+
+```    
+In this example, we are exporting two functions: `getOrderByID` and `getOrderStatus`. Other applications or modules can use these functions by importing the module as explained in the next section.
+
+## Importing Blocks of Reusable Code  
+We can import one or more modules into other modules or applications which want to use the functions defined in those modules. 
+
+Let us import the module created in the previous section in another module: `OrderRoutes.js` by using the `require` function:
+```js
+const express = require('express')
+const router = express.Router()
+
+
+// Import the OrderInquiryController module 
+const  { 
+    getOrders,getOrderByID,getOrderStatus
+} = require('../controllers/OrderInquiryController.js')
+
+router.get('/', getOrders)
+router.get('/:orderId', getOrderByID)
+router.post('/:orderId/status', getOrderStatus)
+```
+In this code snippet, we have imported the module: `OrderInquiryController`. We have used a relative path: `../controllers/OrderInquiryController.js` to specify the location of the module.
+
+We can also publish modules in a shared module registry, and other applications or modules can use them by installing from the shared module registry using the npm package manager. These installed modules reside in the `node_modules` folder.
+
+## Applying the Principle of Separation of Concerns for Organizing Code
+Separation of concerns is a principle of software design used to break down an application into independent units with minimal overlap between the functions of the individual units.
+In Node.js, we can separate our code into different files and directories based on their functionality. 
+
+For example, we can keep all our controllers in a `controllers` directory, and all your routes in a `routes` directory. This approach makes it easier to locate specific pieces of logic in a huge codebase thereby making the code readable and maintainable.
+
+This is an example of grouping files and folders using the principle of Separation of Concerns by roles:
+
+```shell
+│   ├── app.js
+│       ├── controllers
+│       │   ├── inquiryController.js
+│       │   └── updateController.js
+│       ├── dbaccessors
+│       │   └── dataAccessor.js
+│       ├── models
+│       │   └── order.js
+│       ├── routes
+│       │   └── routes.js
+│       └── services
+│           └── inquiryService.js
+```
+As we can see, the controller files: `OrderInquiryController.js` and `OrderUpdateController.js` are in one folder: `controllers`. Similarly, we have created folders for putting other types of files like `routes`, `models`, `services`, and `dbaccessors`. 
+
+This method of grouping by roles should be used for smaller codebases typically less than 4 features. For more than 4 features we should organize by features rather than by roles as explained in the next section.
```

Comment:
Consider expanding on the explanation of why organizing by features is recommended for codebases with more than 4 features.

Reply:
done